File: LeetCode_Algorithm/1021(Remove Outermost Parentheses)/1021.py
'''
# 가장 바깥에 존재하는 괄호 없애기
# sol1 42ms 81% 16.1MB 99%
O(n) O(1)

'''

class Solution:
    def removeOuterParentheses(self, s: str) -> str:
        # sol
        cnt = 0; ans = ""; l = 0
        for r in range(len(s)):
            if s[r] == '(':
                cnt += 1;
            else:
                cnt -= 1
            if cnt == 0:
                ans += s[l+1: r]
                l = r+1
        return ans
print(Solution().removeOuterParentheses("(()())(())"))
print(Solution().removeOuterParentheses("(()())(())(()(()))"))

File: LeetCode_Algorithm/1021(Remove Outermost Parentheses)/1021.js
var removeOuterParentheses = function(s) {
  // sol1 55ms 82% 42MB 98%
  let cnt = 0; let l = 0; let ans = ""
  for (let r = 0; r < s.length; r++) {
    s[r] == '(' ? cnt += 1 : cnt -= 1
    if (cnt == 0) {
      ans += s.slice(l+1, r)
      l = r + 1
    }
  }  
  return ans
};
console.log(removeOuterParentheses("(()())(())"))